File: python/src/chatgpt_oauth/pkce.py
"""PKCE generation, base64url helpers, and constant-work state comparison."""

import base64
import hashlib
import secrets

from .types import StateMismatchError


def base64url(value: bytes) -> str:
    """Encodes bytes as unpadded RFC 4648 URL-safe base64."""
    return base64.urlsafe_b64encode(value).rstrip(b"=").decode("ascii")


def decode_base64url(value: str) -> bytes:
    """Decodes unpadded RFC 4648 URL-safe base64."""
    return base64.urlsafe_b64decode(value + "=" * (-len(value) % 4))


def pkce_challenge(verifier: str) -> str:
    """Computes an OAuth S256 code challenge."""
    return base64url(hashlib.sha256(verifier.encode()).digest())


def create_pkce() -> tuple[str, str, str]:
    """Creates a verifier, its challenge, and independent callback state."""
    verifier = base64url(secrets.token_bytes(64))
    return verifier, pkce_challenge(verifier), base64url(secrets.token_bytes(32))


def assert_state(expected: str, actual: str | None) -> None:
    """Rejects state mismatch without a data-dependent early exit."""
    left = expected.encode()
    right = (actual or "").encode()
    difference = len(left) ^ len(right)
    for index in range(max(len(left), len(right))):
        difference |= (left[index] if index < len(left) else 0) ^ (
            right[index] if index < len(right) else 0
        )
    if difference:
        raise StateMismatchError

File: kotlin/chatgpt-oauth-core/src/main/kotlin/dev/chatgptoauth/Pkce.kt
package dev.chatgptoauth

import java.nio.charset.StandardCharsets
import java.security.MessageDigest
import java.security.SecureRandom
import java.util.Base64

/**
 * Holds protected PKCE material for one short-lived authorization attempt.
 *
 * @property verifier Secret proof retained until code exchange.
 * @property challenge Public S256 digest sent to authorization.
 * @property state Secret callback correlation value.
 */
internal data class PkceData(
    val verifier: String,
    val challenge: String,
    val state: String,
) {
    override fun toString(): String = "PkceData(verifier=[REDACTED], challenge=$challenge, state=[REDACTED])"
}

/** Encodes bytes as RFC 4648 URL-safe Base64 without padding. */
internal fun base64Url(bytes: ByteArray): String = Base64.getUrlEncoder().withoutPadding().encodeToString(bytes)

/** Decodes RFC 4648 URL-safe Base64 with omitted padding accepted. */
internal fun decodeBase64Url(value: String): ByteArray = Base64.getUrlDecoder().decode(value)

/** Computes the S256 PKCE challenge for a verifier. */
internal fun pkceChallenge(verifier: String): String = base64Url(
    MessageDigest.getInstance("SHA-256").digest(verifier.toByteArray(StandardCharsets.UTF_8)),
)

/** Generates a fresh verifier, challenge, and state with a cryptographic RNG. */
internal fun createPkce(random: SecureRandom = SecureRandom()): PkceData {
    val verifierBytes = ByteArray(64).also(random::nextBytes)
    val stateBytes = ByteArray(32).also(random::nextBytes)
    val verifier = base64Url(verifierBytes)
    return PkceData(verifier, pkceChallenge(verifier), base64Url(stateBytes))
}

/** Validates callback state without a data-dependent early exit. */
internal fun assertState(expected: String, actual: String?) {
    val left = expected.toByteArray(StandardCharsets.UTF_8)
    val right = (actual ?: "").toByteArray(StandardCharsets.UTF_8)
    var difference = left.size xor right.size
    repeat(maxOf(left.size, right.size)) { index ->
        difference = difference or ((left.getOrElse(index) { 0 }).toInt() xor (right.getOrElse(index) { 0 }).toInt())
    }
    if (difference != 0) throw StateMismatchError()
}
